Ttva 105 securitas integration

danjacob-anders opened the pull request.

danjacob-anders committed:
Securitas: add Kulkunen implementation

Refs: TTVA-105

danjacob-anders committed:
Kulkunen: reformat models.py

Module formatted with isort and black.

Refs: TTVA-105

danjacob-anders committed:
Securitas: add Securitas Driver to access options

Refs: TTVA-105

danjacob-anders committed:
Securitas: add unit tests

Refs: TTVA-105

danjacob-anders committed:
Kulkunen: add users and grants to Django admin

This also tidies up existing admin classes.

Refs: ttva-105-securitas-integration

danjacob-anders committed:
Resource: add ResourceAccess model

In Django or Respa admin one or more access methods can be added to a
Resource.

Refs: TTVA-105

danjacob-anders committed:
Resources: add access method to admin and API

Refs: TTVA-105

danjacob-anders committed:
Respa admin: add access methods to Resource admin form

Refs: TTVA-105

danjacob-anders committed:
Resources: add data migration for default access methods

Refs: TTVA-105

danjacob-anders committed:
Kulkunen: add better error logging for grant removal

logger.exception() should pass the Exception instance

Refs: TTVA-105

danjacob-anders committed:
Resources: add merge migration

This migration includes merge of SAP and Kulkunen migrations.

Refs: TTVA-105

danjacob-anders commented on kulkunen/admin.py:
```diff
@@ -1,28 +1,104 @@
-    list_display = ('resource', 'system', 'driver_identifier', 'active_grant_count')
-
-    def get_form(self, request, obj=None, **kwargs):
-        schema = {}
```
Removed the JSONEditorWidget as it was not working properly and made it hard to edit the JSON in the Django admin.

danjacob-anders commented on kulkunen/drivers/securitas.py:
```diff
@@ -0,0 +1,173 @@
+    RESOURCE_CONFIG_SCHEMA = {
+        "type": "object",
+        "properties": {
+            "uses_pincode": {
```
This setting, "uses_pincode" can be set per-resource, with default being true. Some resources will not be accessible by pincode, just mobile app.

danjacob-anders commented on kulkunen/drivers/securitas.py:
```diff
@@ -0,0 +1,173 @@
+        "type": "object",
+        "properties": {
+            "api_key": {
+                "type": "string",
```
API key from Securitas, set in Django admin for the system.

danjacob-anders commented on kulkunen/drivers/securitas.py:
```diff
@@ -0,0 +1,173 @@
+            data = response.json()
+            grant.access_code = data["code"]
+            # save the Securitas ID to driver data so we can delete later
+            grant.driver_data = {"code_id": data["id"]}
```
"code_id" is the primary key on Securitas' side, distinct from the actual pincode.

danjacob-anders commented on kulkunen/drivers/securitas.py:
```diff
@@ -0,0 +1,173 @@
+            response.raise_for_status()
+            return response
+        except requests.RequestException as e:
+            print(params)
```
These print statements can probably be removed now, used for debugging in QA.

danjacob-anders commented on kulkunen/drivers/securitas.py:
```diff
@@ -0,0 +1,173 @@
+            response.raise_for_status()
+            return response
+        except requests.RequestException as e:
+            print(e)
```
These print statements can probably be removed now, used for debugging in QA.

danjacob-anders commented on resources/models/resource.py:
```diff
@@ -241,6 +241,21 @@ def free_of_charge(self, is_free):
 
 
+class ResourceAccess(models.Model):
+    ACCESS_METHODS = (
```
Access method shown on Respa admin/Varaamo.

This is not handled dynamically, as we cannot know necessarily from the Kulkunen system which locking system is used or even if there is a Kulkunen system, so the admin should set these in the Respa admin.

Because this is M:M (there can be multiple methods for a Resource,e.g. Pincode+Mobile App) I used a model here rather than a simple CharField+choices.

jopesy commented on locale/fi/LC_MESSAGES/django.po:
```diff
@@ -1102,6 +1102,24 @@ msgstr "Vahva"
+msgstr "Perinteinen avain/tunnistelätkä"
+
+msgid "Security guard/janitor"
+msgstr "Vahtimestari/Valvooja"
```
```suggestion
msgstr "Vahtimestari/Valvoja"
```

danjacob-anders committed:
Securitas: use resource-group-access endpoint instead of access endpoint
to obtain mobile access for resource

Refs: TTVA-105

jopesy commented on locale/fi/LC_MESSAGES/django.po:
```diff
@@ -1102,6 +1102,21 @@ msgstr "Vahva"
+
+msgid "Door open"
+msgstr "Ovi auki"
+
```
These should be added to the translations too:
- "Access methods" = "Pääsytavat"
- "Resource access method" = "Resurssin pääsytapa"
- "Resource access methods" = "Resurssin pääsytavat"

jopesy commented on resources/migrations/0120_merge_20231117_1306.py:
```diff
@@ -0,0 +1,12 @@
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
```
You could also delete the migrations 0117 and 0118 added by this PR (after reverting them locally of course)  and recreate them. Then merge migrations wouldn't be needed.

danjacob-anders committed:
Securitas: Ensure resource has group_id in metadata before adding and
removing grant.

Additional tests added to check for the group_id. Note that Kulkunen
does have validation endpoints, but these do not appear to be called by
the management command or elsewhere e.g. Django admin. A future addition
should integrate these into the Django admin to prevent runtime issues
with misconfiguration.

Refs: TTVA-105

jopesy reviewed (approved):
Looks really good! 💯
